services/notifier/notifier.go

```diff
@@ -0,0 +1,496 @@
+	logger              logger.Logger
+	statsFactory        stats.Stats
+	db                  *sqlmw.DB
+	repo                *repo.Notifier
+	workspaceIdentifier string
+	batchIDGenerator    func() uuid.UUID
+	now                 func() time.Time
+	background          struct {
+		group  *errgroup.Group
+		ctx    context.Context
+		cancel context.CancelFunc
+		wait   func() error
+	}
+
+	config struct {
+		host                       string
+		port                       int
+		user                       string
+		password                   string
+		database                   string
+		sslMode                    string
+		maxAttempt                 int
+		maxOpenConnections         int
+		shouldForceSetLowerVersion bool
+		trackBatchInterval         time.Duration
+		maxPollSleep               time.Duration
+		jobOrphanTimeout           time.Duration
+		queryTimeout               time.Duration
+	}
+	stats struct {
+		insertRecords      stats.Measurement
+		publish            stats.Measurement
+		publishTime        stats.Measurement
+		claimLag           stats.Measurement
+		trackBatchLag      stats.Measurement
+		claimSucceeded     stats.Measurement
+		claimSucceededTime stats.Measurement
+		claimFailed        stats.Measurement
+		claimFailedTime    stats.Measurement
+		claimUpdateFailed  stats.Measurement
+		abortedRecords     stats.Measurement
+	}
+}
+
+func New(
+	ctx context.Context,
+	conf *config.Config,
+	log logger.Logger,
+	statsFactory stats.Stats,
+	workspaceIdentifier string,
+	fallbackDSN string,
+) (*Notifier, error) {
+	n := &Notifier{
+		conf:                conf,
+		logger:              log.Child("notifier"),
+		statsFactory:        statsFactory,
+		workspaceIdentifier: workspaceIdentifier,
+		batchIDGenerator:    misc.FastUUID,
+		now:                 time.Now,
+	}
+
+	n.logger.Infof("Initializing Notifier...")
+
+	n.config.host = n.conf.GetString("PGNOTIFIER_DB_HOST", "localhost")
+	n.config.user = n.conf.GetString("PGNOTIFIER_DB_USER", "ubuntu")
+	n.config.database = n.conf.GetString("PGNOTIFIER_DB_NAME", "ubuntu")
+	n.config.port = n.conf.GetInt("PGNOTIFIER_DB_PORT", 5432)
+	n.config.password = n.conf.GetString("PGNOTIFIER_DB_PASSWORD", "ubuntu")
+	n.config.sslMode = n.conf.GetString("PGNOTIFIER_DB_SSL_MODE", "disable")
+	n.config.maxAttempt = n.conf.GetInt("PgNotifier.maxAttempt", 3)
+	n.config.maxOpenConnections = n.conf.GetInt("PgNotifier.maxOpenConnections", 20)
+	n.config.shouldForceSetLowerVersion = n.conf.GetBool("SQLMigrator.forceSetLowerVersion", true)
+	n.config.trackBatchInterval = n.conf.GetDuration("PgNotifier.trackBatchIntervalInS", 2, time.Second)
+	n.config.queryTimeout = n.conf.GetDuration("Warehouse.pgNotifierQueryTimeout", 5, time.Minute)
+
+	n.conf.RegisterDurationConfigVariable(5000, &n.config.maxPollSleep, true, time.Millisecond, "PgNotifier.maxPollSleep")
+	n.conf.RegisterDurationConfigVariable(120, &n.config.jobOrphanTimeout, true, time.Second, "PgNotifier.jobOrphanTimeout")
```

Comment:
Hot reloadables should be atomic or be read when needed.
https://github.com/rudderlabs/rudder-server/pull/3805#discussion_r1311266977